src/rabbit_variable_queue.erl

```diff
@@ -985,16 +985,24 @@ is_duplicate(_Msg, State) -> {false, State}.
 
 set_queue_mode(Mode, State = #vqstate { mode = Mode }) ->
     State;
-set_queue_mode(lazy, State = #vqstate {
-                                target_ram_count = TargetRamCount }) ->
-    %% To become a lazy queue we need to page everything to disk first.
-    State1 = convert_to_lazy(State),
-    %% restore the original target_ram_count
-    a(State1 #vqstate { mode = lazy, target_ram_count = TargetRamCount });
-set_queue_mode(default, State) ->
+set_queue_mode(lazy, State0) ->
+    %% Note: rabbitmq/rabbitmq-server#1284
+    %% We must override the value set by queue_index_embed_msgs_below since
+    %% restoring a queue with embedded small messages will restore those
+    %% messages as betas in RAM, which we do not want with a lazy queue.
+    %% We also explicitly set target_ram_count to 0 to reinforce this.
+    State1 = State0#vqstate { qi_embed_msgs_below = 0,
+                              target_ram_count = 0 },
+    State2 = convert_to_lazy(State1),
```

Comment:
Prior to these changes, after switching to `lazy` mode, `target_ram_count` would be set back to the initial value saved in `TargetRamCount` which would be `infinity`. This value doesn't really make sense in the case of `lazy` queues as we do *not* want to have any messages in RAM.

Setting the value to `0` does not seem to have any ill effects, but I will re-re-review that change and take it into consideration as I test some more scenarios today.